efficientdet/keras/train_lib_test.py

```diff
@@ -0,0 +1,85 @@
+# Lint as: python3
+# Copyright 2020 Google Research. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from absl.testing import parameterized
+from absl import logging
+
+import tensorflow as tf
+
+import hparams_config
+from keras.train_lib import StepwiseLrSchedule, CosineLrSchedule, PolynomialLrSchedule, BoxLoss, BoxIouLoss, FocalLoss, EfficientDetNetTrain, get_optimizer
+from det_model_fn import stepwise_lr_schedule, cosine_lr_schedule, polynomial_lr_schedule, _box_loss
+
+
+class TrainLibTest(tf.test.TestCase, parameterized.TestCase):
+
+  def test_lr_schedule(self):
+    stepwise = StepwiseLrSchedule(1e-3, 1e-4, 1, 3, 5)
+    cosine = CosineLrSchedule(1e-3, 1e-4, 1, 5)
+    polynomial = PolynomialLrSchedule(1e-3, 1e-4, 1, 2, 5)
+    for i in range(5):
+      self.assertEqual(stepwise_lr_schedule(1e-3, 1e-4, 1, 3, 5, i),
+                       stepwise(i))
+      self.assertEqual(cosine_lr_schedule(1e-3, 1e-4, 1, 5, i), cosine(i))
+      self.assertEqual(polynomial_lr_schedule(1e-3, 1e-4, 1, 2, 5, i),
+                       polynomial(i))
+
+  def test_losses(self):
+    box_loss = BoxLoss()
+    box_outputs = tf.ones([10])
+    box_targets = tf.zeros([10])
+    num_positives = 4.0
+    self.assertEqual(_box_loss(box_outputs, box_targets, num_positives),
+                     box_loss([num_positives, box_targets], box_outputs))
+
+  def test_train(self):
+    config = hparams_config.get_detection_config('efficientdet-d0')
+    config.batch_size = 1
+    config.num_examples_per_epoch = 1
+    x = tf.ones((1, 512, 512, 3))
+    labels = {
+        'box_targets_%d' % i: tf.ones((1, 512 // 2**i, 512 // 2**i, 36))
+        for i in range(3, 8)
+    }
+    labels.update({
+        'cls_targets_%d' % i: tf.ones((1, 512 // 2**i, 512 // 2**i, 9),
+                                      dtype=tf.int32) for i in range(3, 8)
+    })
+    labels.update({'mean_num_positives': 100.0})
+
+    model = EfficientDetNetTrain(config=config)
+    params = config.as_dict()
+    model.compile(optimizer=get_optimizer(params),
+                  loss={
```

Comment:
I still don't understand how loss is computed here.  In the keras doc, it says it would compute all losses in this dict and sum them up.

Here each loss takes some tensors as inputs, but the network outputs and groundtruth are a list of tensors for different levels. How they are connected?

Reply:
I use same logic as origin code.

Reply:
So this loss parameter here is not really useful? (the doc says if you pass in loss, then you can use something like compiled_loss)

Reply:
I don't use self.compiled_loss, I use self.loss instead. It's a dict.

Reply:
I can't monitor box loss and class loss if I use compiled_loss.

Reply:
Gotcha!